kernel/src/lib.rs

```diff
@@ -415,3 +418,70 @@ pub trait Engine: AsAny {
     /// Get the connector provided [`ParquetHandler`].
     fn get_parquet_handler(&self) -> Arc<dyn ParquetHandler>;
 }
+
+#[cfg(doctest)]
+mod doc_tests {
+
+    /// ```
+    /// # use delta_kernel_derive::Schema;
+    /// #[derive(Schema)]
+    /// pub struct WithFields {
+    ///     some_name: String,
+    /// }
+    /// ```
+    #[cfg(doctest)]
+    pub struct MacroTestStructWithField;
+
+    /// ```compile_fail
+    /// # use delta_kernel_derive::Schema;
+    /// #[derive(Schema)]
+    /// pub struct NoFields;
+    /// ```
+    #[cfg(doctest)]
+    pub struct MacroTestStructWithoutField;
+
+    /// ```
+    /// # use delta_kernel_derive::Schema;
+    /// # use std::collections::HashMap;
+    /// #[derive(Schema)]
+    /// pub struct WithAngleBracketPath {
+    ///     map_field: HashMap<String, String>,
+    /// }
+    /// ```
+    #[cfg(doctest)]
+    pub struct MacroTestStructWithAngleBracketedPathField;
+
+    /// ```
+    /// # use delta_kernel_derive::Schema;
+    /// # use std::collections::HashMap;
+    /// #[derive(Schema)]
+    /// pub struct WithAttributedField {
+    ///     #[drop_null_container_values]
+    ///     map_field: HashMap<String, String>,
+    /// }
+    /// ```
+    #[cfg(doctest)]
+    pub struct MacroTestStructWithAttributedField;
+
+    /// ```compile_fail
+    /// # use delta_kernel_derive::Schema;
+    /// #[derive(Schema)]
+    /// pub struct WithInvalidAttributeTarget {
+    ///     #[drop_null_container_values]
+    ///     some_name: String,
+    /// }
+    /// ```
+    #[cfg(doctest)]
+    pub struct MacroTestStructWithInvalidAttributeTarget;
+
+    /// ```compile_fail
+    /// # use delta_kernel_derive::Schema;
+    /// # use syn::Token;
+    /// #[derive(Schema)]
+    /// pub struct WithInvalidFieldType {
+    ///     token: Token![struct],
```

Comment:
It is the latter, I am trying to test all the code paths in the macro. Actually, there is one path I cannot trigger. For that
I need to craft something with `PathArguments::Parenthesized` but I run out of ideas.

I asked around [on StackOverlow here](https://stackoverflow.com/questions/79123128/how-to-declare-synpatharguments-parenthesized-field-in-structure-in-rust]) but didn't get anything solid.

 This one:
```
_ => Error::new(segment.arguments.span(), "Can only handle <> type path args").to_compile_error()
```

Reply:
> > What is `Token![struct]`? Is it intentionally introducing a rust syntax error to see if the macro handles it gracefully? Or does it somehow produce valid rust syntax that we expect the macro cannot handle?
>
> It is the latter, I am trying to test all the code paths in the macro. 

Maybe silly question, but: If it's valid rust syntax, why can't we just type it in directly? Why do we need to inject it with syn?

Reply:
Typing it directly will not trigger the compile error in the macro I was trying to trigger (it will create another compile error but not the one I am trying to cause), this line in the `derive-macros/src/lib.rs`
```
_ => Error::new(field.span(), format!("Can't handle type: {:?}", field.ty)).to_compile_error()
```

Token! was the first type I found that I found which would help me trigger that and therefore I used it. Now I just realised that `syn` is actually not a `dev-dependency` of the project so it has to be added it to `Cargo.toml`. However, I'll try to see if I can find a type which I can use without adding `syn`

Reply:
This is actually a reason why the previous version, one using `trybuild`, was a bit better IMO as it would capture the exact error whereas here I'm only causing the compile error without know exactly which one is raised

Reply:
Yes... but the downside is every rustc update has the potential to change the error text, ie https://github.com/delta-io/delta-kernel-rs/pull/346 (the FFI `Handle` API currently relies on trybuild tests, because I hadn't learned the export+visibility tricks that would allow doc tests to work). 

On the other hand, codecov doesn't seem to "count" executable doctests as coverage for some reason, while presumably it would count trybuild tests, to the extent they exercise any kernel code. Maybe that's just a problem with our CI, I don't know?

Reply:
Totally agree, neither solution is ideal imo but the one with doctests requires less effort to create and maintain so I'll stick with that (this PR) but will stick to `compile_fail` scenarios only (maybe have one test that succeeds just to assure the methodology is still working)